### DISCUSSION TITLE
.NET 9 Preview 4

### DISCUSSION THREAD

**Discussion by jamesmontemagno** [dotnet/core#9318] [News]
We are pleased to announce the release of .NET 9 Preview 4. This release contains the following Libraries & Runtime improvements:

- [`UnsafeAccessorAttribute` supports generic parameters](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/runtime.md#unsafeaccessorattribute-supports-generic-parameters)
- [Feature switches with trimming support](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/runtime.md#feature-switches-with-trimming-support)
- [Tokenizer library enhancements](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/libraries.md#tokenizer-library-enhancements)
- [OpenTelemetry: Make activity linking more flexible](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/libraries.md#opentelemetry-make-activity-linking-more-flexible)
- [PDB support for System.Reflection.Emit.PersistedAssemblyBuilder](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/libraries.md#pdb-support-added-for-systemreflectionemitpersistedassemblybuilder)

For details on these new features and more see the [Libraries release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/libraries.md) and [Runtime release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/runtime.md)

Discuss this release with the .NET team on [GitHub Discussions](https://github.com/dotnet/runtime/discussions/102504).

You can find detailed release notes for additional features in .NET 9 Preview 4 below:

###  .NET Data & EF Core

- [Enhanced database provider for Azure Cosmos DB for NoSQL](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/efcoreanddata.md#enhanced-database-provider-for-azure-cosmos-db-for-nosql)
- [GroupBy complex types](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/efcoreanddata.md#groupby-complex-types)
- [Read-only primitive collections](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/efcoreanddata.md#read-only-primitive-collections)
- Full [release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/efcoreanddata.md)

### .NET MAUI

- [New Input Keyboards](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/dotnetmaui.md#new-input-keyboards)
- [Android 15 Beta 1 Support](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/dotnetmaui.md#android-15-beta-1-support)
- [NativeAOT improvements for iOS & Mac Catalyst](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/dotnetmaui.md#native-aot-for-ios--mac-catalyst)
- Full [release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/dotnetmaui.md)

Discuss this release with the .NET MAUI team on [GitHub Discussions](https://github.com/dotnet/maui/discussions/22553)

### ASP.NET Core

- [Built-in support for OpenAPI document generation](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/aspnetcore.md#built-in-support-for-openapi-document-generation)
- [Introducing `HybridCache`](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/aspnetcore.md#introducing-hybridcache)
- [Developer exception page improvements](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/aspnetcore.md#developer-exception-page-improvements)
- [Added `OverscanCount` parameter to `QuickGrid`](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/aspnetcore.md#added-overscancount-parameter-to-quickgrid)
- [Add static SSR pages to a globally-interactive Blazor Web app](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/aspnetcore.md#add-static-ssr-pages-to-a-globally-interactive-blazor-web-app)
- [Avoid 503's during app recycle in IIS](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/aspnetcore.md#avoid-503s-during-app-recycle-in-iis)
- Full [release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/aspnetcore.md)

Discuss this release with the ASP.NET Core team on [GitHub Discussions](https://github.com/dotnet/aspnetcore/discussions/55820).

### WPF

- [Theming improvements – Support for Windows 11 theming for WPF apps](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/wpf.md#support-for-windows-11-theming)
- [Hyphen based ligature support](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/wpf.md#hyphen-based-ligature-support)
- Full [release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/wpf.md)

## Get started
To get started with .NET 9, [install the .NET 9 SDK](https://dotnet.microsoft.com/download/dotnet/9.0).

If you're on Windows using Visual Studio, we recommend installing the latest [Visual Studio 2022 preview](https://visualstudio.microsoft.com/vs/preview/), or get started with Visual Studio Code and the [C# Dev Kit](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.csdevkit) extension.

You can find [.NET 9 Preview 4 Installer & Binaries](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/9.0.0-preview.4.md) on GitHub.

## Team Announcements & Discussions
Discuss this release with the product teams on GitHub through the announcements for this release. 

* [.NET MAUI](https://github.com/dotnet/maui/discussions/22553)
* [ASP.NET Core](https://github.com/dotnet/aspnetcore/discussions/55820)
* [Libraries & Runtime](https://github.com/dotnet/runtime/discussions/102504)
* [Source-build](https://github.com/dotnet/source-build/discussions/4414)

## Stay up-to-date with .NET 9
You can stay up-to-date with all the features of .NET 9 with:

* [What's new in .NET 9](https://learn.microsoft.com/dotnet/core/whats-new/dotnet-9/overview)
* [What's new in ASP.NET Core](https://learn.microsoft.com/aspnet/core/release-notes/aspnetcore-9.0)
* [What's new in .NET MAUI](https://learn.microsoft.com/dotnet/maui/whats-new/dotnet-9)
* [What's new in EF Core](https://learn.microsoft.com/ef/core/what-is-new/ef-core-9.0/whatsnew)
* [Breaking Changes in .NET 9](https://learn.microsoft.com/dotnet/core/compatibility/9.0)
* [.NET 9 Releases](https://github.com/dotnet/core/blob/main/release-notes/9.0/README.md)

For .NET 9 previews we are utilizing GitHub Discussions to make announcements. For more information [read our discussion](https://github.com/dotnet/core/discussions/9131).

Stay up to date by subscribing to the [RSS feed](https://github.com/dotnet/core/discussions/categories/news.atom) for all release announcements.

Engage with the product teams by commenting below as you try out these new features or have questions.

**Comment by hez2010** [comment]
The link to MAUI release notes is incorrect. It should be https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/dotnetmaui.md

**Reply by jamesmontemagno** [reply]
This is updated

**Comment by jeromelaban** [comment]
Congrats on the release!

One thing `dotnet workload install maui` fails with:

```
Workload installation failed: One or more errors occurred. (Version 17.0.8523 of package microsoft.ios.sdk.net8.0_17.0.msi.x64 is not found in NuGet feeds https://api.nuget.org/v3/index.json.)
```

Should we be waiting for some package publication later today? I know @rolfbjarne had to do something for preview 3, maybe it needs to be done the same for preview 4?

**Reply by rolfbjarne** [reply]
Hm yeah, looks like some packages are missing. Looking into it now.

**Reply by rolfbjarne** [reply]
Should be all published now.

**Reply by jeromelaban** [reply]
It's working, thanks for the quick update!

**Comment by mhedden94** [comment]
Glad to see support for WPF and Windows 11 styling.  However, even after installing the latest .NET 9 preview 4 SDK and preview version of Visual Studio 2022, building a WPF app on Windows 11 still shows the old control design.  I don’t see any option to enable or force the new design elements.

**Reply by hez2010** [reply]
Adding below lines to your App.xaml to enable the new Windows 11 fluent design:
```xaml
<Application.Resources>
    <ResourceDictionary>
        <ResourceDictionary.MergedDictionaries>
            <ResourceDictionary Source="pack://application:,,,/PresentationFramework.Fluent;component/Resources/Fluent.xaml" />
        </ResourceDictionary.MergedDictionaries>
    </ResourceDictionary>
</Application.Resources>
```

**Reply by mhedden94** [reply]
Worked perfectly, thank you!

**Reply by jamesmontemagno** [reply]
Updated the docs with this :)

**Reply by kumoh2** [reply]
nice

**Reply by arknu** [reply]
@jamesmontemagno Is this intended to be needed for the final release? If so, seems kind of silly. Who would not want the new styles by default? You could have a switch the other way round, if you need the old styles for compatibility.

**Reply by jamesmontemagno** [reply]
I would inquire on the WPF repo: https://github.com/dotnet/wpf

**Comment by PaskalSunari** [comment]
Congrats for the release

**Comment by IrinaPykhova** [comment]
Congratulations with release. @jamesmontemagno I have some questions about WPF Fluent theme:

- if you use it in application, what would happen if this application runs on Windows 10?
- what are your plans about target frameworks? Would it be .Net 9 only, or we will be able to use it in earlier versions? What about .Net framework? 
- do you plan something for control authoring? For old themes, you could include theme-specific resources into your library as described here: https://learn.microsoft.com/en-us/dotnet/desktop/wpf/controls/control-authoring-overview?view=netframeworkdesktop-4.8#defining-resources-at-the-theme-level. Would it be possible to do the same for Fluent theme? Or maybe you will obslete this mechanism entirely?

**Reply by IrinaPykhova** [reply]
@jamesmontemagno do you have any insights about my questions?

**Reply by jamesmontemagno** [reply]
Best thing to do would be to ask on the WPF repo: https://github.com/dotnet/wpf

**Comment by valco1994** [comment]
I can't compile .NET 9 application on macOS Big Sur x64 with .NET 9 preview 4 due to `Symbol not found: __ZNSt3__113basic_filebufIcNS_11char_traitsIcEEE4openEPKcj` in libhostfxr.dylib. 

Is it intentional? Does .NET 9 require macOS 12+ now?

**Reply by jamesmontemagno** [reply]
@richlander ?

**Reply by richlander** [reply]
@elinor-fung ?

**Reply by AlexanderK46** [reply]
See https://github.com/dotnet/sdk/issues/45382

**Comment by fraluderin** [comment]
Congratulations on this release and thanks to all the contributors for this great result... finally WPF gets the attention it deserves! I can't believe it is finally getting theme updates. Any news on extending nullable context annotations on WPF library?
